File: scala/sdktest/EcomParameterEntityTest.scala
// Generated basic-flow test for the ecom_parameter entity (model-driven;
// mirrors the java TestEntity generator). A dependency-free scala-cli test
// object driven by SdkEntityTestMain. Runs against the in-memory test
// transport seeded with the shipped EcomParameterTestData.json fixtures.

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}

import voxgig.bluefintecsmerchantservicessdk.core.{Helpers, BluefinTecsMerchantServicesSDK}
import voxgig.bluefintecsmerchantservicessdk.utility.struct.Struct

object EcomParameterEntityTest {

  def run(rep: SdkTestReport): Unit = {
    rep.scope("ecom_parameter.instance") {
      val testsdk = BluefinTecsMerchantServicesSDK.testSDK()
      val ent = testsdk.ecomParameter(null)
      rep.check("ecom_parameter.instance", ent != null, "expected non-null ecom_parameter entity")
    }

    rep.scope("ecom_parameter.basic") {
      val entityData = Helpers.toMapAny(SdkTestSupport.readJson(
          "../.sdk/test/entity/ecom_parameter/EcomParameterTestData.json"))
      val options = new LinkedHashMap[String, Object]()
      options.put("entity", entityData.get("existing"))
      val client = BluefinTecsMerchantServicesSDK.testSDK(options, null)

      val idmap = new LinkedHashMap[String, Object]()
      idmap.put("ecom_parameter01", "ECOM_PARAMETER01")
      idmap.put("ecom_parameter02", "ECOM_PARAMETER02")
      idmap.put("ecom_parameter03", "ECOM_PARAMETER03")
      val now = System.currentTimeMillis()

      // CREATE
      val ecomParameterRef01Ent = client.ecomParameter(null)
      var ecomParameterRef01Data = Helpers.toMapAny(Struct.getprop(
          Struct.getpath(entityData, "new.ecom_parameter"), "ecom_parameter_ref01"))
      val ecomParameterRef01DataResult = ecomParameterRef01Ent.create(ecomParameterRef01Data, null)
      ecomParameterRef01Data = Helpers.toMapAny(ecomParameterRef01DataResult)
      rep.check("ecom_parameter.create.map", ecomParameterRef01Data != null, "expected create result to be a map")
    }
  }
}

File: cpp/test/ecom_parameter_entity_test.cpp
// Generated basic-flow test for the ecom_parameter entity (model-driven,
// unit mode; mirrors the rust/go TestEntity generator).

#include "runner_support.hpp"

using namespace sdk;
using namespace sdk::rs;

struct EcomParameterSetup {
  std::shared_ptr<BluefinTecsMerchantServicesSDK> client;
  Value data;
  Value idmap;
  Value env;
  bool live = false;
  bool synthetic_only = false;
  long long now = 0;
};

static EcomParameterSetup ecom_parameter_basic_setup(const Value& extra) {
  load_env_local();

  std::string entity_data_file = "../.sdk/test/entity/ecom_parameter/EcomParameterTestData.json";
  Value entity_data = vs::parse_json(read_file(entity_data_file));

  Value options = vmap({{"entity", getp(entity_data, "existing")}});
  auto client = BluefinTecsMerchantServicesSDK::testSDK(options, extra);

  // idmap via transform (upper-cased id name synthetics), matching the donors.
  Value idmap = Struct::transform(
      vlist({Value("ecom_parameter01"), Value("ecom_parameter02"), Value("ecom_parameter03")}),
      vmap({{"`$PACK`", vlist({
        Value(""),
        vmap({
          {"`$KEY`", Value("`$COPY`")},
          {"`$VAL`", vlist({Value("`$FORMAT`"), Value("upper"), Value("`$COPY`")})}
        })
      })}}));
  if (!idmap.is_map()) idmap = vmap();

  Value env = env_override(vmap({
    {"BLUEFIN_TECS_MERCHANT_SERVICES_TEST_ECOM_PARAMETER_ENTID", idmap},
    {"BLUEFIN_TECS_MERCHANT_SERVICES_TEST_LIVE", Value("FALSE")},
    {"BLUEFIN_TECS_MERCHANT_SERVICES_TEST_EXPLAIN", Value("FALSE")}
  }));

  Value idmap_resolved = Helpers::toMapAny(getp(env, "BLUEFIN_TECS_MERCHANT_SERVICES_TEST_ECOM_PARAMETER_ENTID"));
  if (!idmap_resolved.is_map()) idmap_resolved = idmap;

  bool live = getp(env, "BLUEFIN_TECS_MERCHANT_SERVICES_TEST_LIVE") == Value("TRUE");

  EcomParameterSetup s;
  s.client = client;
  s.data = entity_data;
  s.idmap = idmap_resolved;
  s.env = env;
  s.live = live;
  s.synthetic_only = false;
  s.now = now_ms();
  return s;
}

static void ecom_parameter_entity_instance() {
  auto testsdk = BluefinTecsMerchantServicesSDK::testSDK();
  auto ent = testsdk->ecom_parameter();
  ASSERT_EQ(ent->getName(), std::string("ecom_parameter"), "entity name");
}


static void ecom_parameter_entity_basic() {
  auto setup = ecom_parameter_basic_setup(Value::undef());
  std::string mode = setup.live ? "live" : "unit";
  for (const std::string& op : std::vector<std::string>{"create"}) {
    auto sk = is_control_skipped("entityOp", std::string("ecom_parameter.") + op, mode);
    if (sk.first) { std::cerr << "skip: " << (sk.second.empty()? "sdk-test-control.json" : sk.second) << "\n"; return; }
  }
  auto client = setup.client;
  // CREATE
  auto ecom_parameter_ref01_ent = client->ecom_parameter();
  Value ecom_parameter_ref01_data = Helpers::toMapAny(getp(Struct::getpath(setup.data, {"new", "ecom_parameter"}), "ecom_parameter_ref01"));
  if (!ecom_parameter_ref01_data.is_map()) ecom_parameter_ref01_data = vmap();
  {
    Value ecom_parameter_ref01_data_result = ecom_parameter_ref01_ent->create(Struct::clone(ecom_parameter_ref01_data), Value::undef())->data();
    ecom_parameter_ref01_data = Helpers::toMapAny(ecom_parameter_ref01_data_result);
    if (!ecom_parameter_ref01_data.is_map()) ecom_parameter_ref01_data = vmap();
    ASSERT_TRUE(ecom_parameter_ref01_data.is_map(), "expected create result to be a map");
  }

}

int main() {
  T_RUN(ecom_parameter_entity_instance);
  T_RUN(ecom_parameter_entity_basic);
  return sdktest::summary("ecom_parameter_entity_test");
}
